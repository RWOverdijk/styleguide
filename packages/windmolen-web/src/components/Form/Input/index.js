// @flow
import React, { type Node, Component } from 'react';
import styled from 'styled-components';
import Autocomplete from 'react-autocomplete';
import { media } from 'styled-bootstrap-grid';
import Base from '../../Base';
import Icon from '../../Icon';
import { colors } from '../../../globals';

type InputProps = {
  error?: Node,
  touched?: boolean,
  isValid?: boolean,
  placeholderRight?: boolean,
  label?: Node,
  onIconClick?: Function,
  type?: string,
  suggestions?: Array<Object>,
  onChange?: Function,
  autoCompleteProps?: Object,

  /** The name of the icon. */
  icon?: string,

  /** CSS class for the container around `<input />`. */
  className?: string
};

const getInputState = (props: InputProps): string => {
  if (props.isValid) {
    return colors.green;
  }

  if (props.error || props.isValid === false) {
    return colors.red;
  }

  return colors.charcoalGray;
};

const StyledInputWrapper = styled(Base.withComponent('div'))`
  position: relative;
`;

const StyledInputLine = styled(Base.withComponent('div'))`
  transition-timing-function: ease-out;
  transition-duration: 0.2s;
  transition-property: background, width;
  bottom: 0;
  left: 50%;
  pointer-events: none;
  height: 1px;
  z-index: 2;
  position: absolute;
  transform: translateX(-50%);
  width: 0;
  background-color: ${colors.charcoalGray};
  ${props => {
    if (props.disabled) {
      return;
    }

    if (props.touched && (!props.isValid || props.error)) {
      return `
        background-color: ${getInputState(props)};
        width: 100%;
      `;
    } else if (props.isValid) {
      return `
        width: 100%;
        background-color: ${getInputState(props)};
      `;
    }
  }}
`;

const StyledInputElement = styled(Base.withComponent('input'))`
  background-color: ${colors.alabaster};
  border: 0;
  box-shadow: inset 0 0 5px 0 rgba(0, 0, 0, 0.12);
  color: ${props => props.disabled ? colors.silver : colors.charcoalGray};
  display: block;
  outline: 0;
  padding: 9px 20px;
  padding-right: ${props => props.icon ? '70px' : '20px'};
  width: 100%;
  height: 50px;
  position: relative;
  z-index: 2;

  &:active,
  &:focus {
    outline: 0;

    & + ${StyledInputLine} {
      ${props => (!props.suggestions && !props.disabled) && `
        background-color: ${getInputState(props)};
        width: 100%;
      `}
    }
  }

  &::placeholder {
    color: ${props => props.disabled ? colors.silver : colors.warmGray};
    text-align: ${props => props.placeholderRight ? 'right' : 'left'};
  }
`;

class StyledInput extends Component<InputProps, { value: string }> {
  input: ?mixed;
  onChange: () => void;

  constructor(props: InputProps) {
    super(props);

    this.state = { value: '' };

    this.input = React.createRef();
    this.onChange = this.onChange.bind(this);
  }

  onChange(event: SyntheticInputEvent<HTMLInputElement>) {
    const { value } = event.target;

    if (typeof this.props.onChange === 'function') {
      this.props.onChange(event);
    }

    this.setState({ value });
  }

  render() {
    // don't let props.onChange overwrite this.onChange
    //
    // eslint-disable-next-line no-unused-vars
    const { onChange, ...props } = this.props;

    return (
      <StyledInputWrapper>
        <StyledInputElement
          innerRef={this.input}
          value={this.state.value}
          onChange={this.onChange}
          {...props}
        />
        <StyledInputLine value={this.state.value} {...this.props} />
      </StyledInputWrapper>
    );
  }
}

const Container = styled(Base)`
  position: relative;
`;

const StyledLabel = styled(Base)`
  color: ${colors.charcoalGray};
  margin-bottom: 5px;

  ${media.desktop`
    margin-bottom: 10px;
  `}
`;

const StyledErrorMessage = styled(Base)`
  font-size: 14px;
  color: ${colors.red};
  line-height: 2;
`;

const StyledIcon = styled(Icon)`
  position: absolute;
  right: 5px;
  top: 5px;
`;

const StyledIconContainer = styled(Base)`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  width: 50px;
  background-color: ${colors.charcoalGray};
  z-index: 3;

  &:focus,
  &:active,
  &:hover {
    background-color: ${colors.shuttleGray};
  }

  &:hover {
    cursor: pointer;
  }
`;

const StyledAutoSuggestion = styled(Base)`
  box-shadow: 0 11px 15px 0 rgba(0, 0, 0, 0.1);
  left: 0;
  padding-bottom: 5px;
  padding-top: 55px;
  position: absolute;
  right: 0;
  top: 0;
  z-index: 1;

  &&& {
    background-color: ${colors.white};
  }

  .suggestion {
    padding-left: 20px;
    padding-right: 20px;
    margin-top: 5px;
    margin-bottom: 5px;
    font-size: 16px;
    line-height: 32px;

    &:focus,
    &:active,
    &:hover {
      color: ${colors.silver};
    }

    &:hover {
      cursor: pointer;
    }
  }
`;

const Input = ({ className, autoCompleteProps, ...props }: InputProps) => {
  return (
    <Container className={className}>
      {props.label && (
        <StyledLabel>{props.label}</StyledLabel>
      )}

      {props.suggestions ? (
        <Autocomplete
          items={props.suggestions}
          wrapperStyle={{  }}
          renderInput={({ ref, value, ...inputProps }) => ( // eslint-disable-line no-unused-vars
            <StyledInput
              innerRef={(node) => ref(node)}
              {...inputProps}
              {...props}
            />
          )}
          renderMenu={(children) => (
            <StyledAutoSuggestion>
              {children}
            </StyledAutoSuggestion>
          )}
          {...autoCompleteProps}
        />
      ) : (
        <StyledInput {...props} />
      )}

      {props.error && (
        <StyledErrorMessage>{props.error}</StyledErrorMessage>
      )}

      {props.icon && (
        <StyledIconContainer onClick={props.onIconClick}>
          <StyledIcon variant={1} name={props.icon} fontSize="40px" />
        </StyledIconContainer>
      )}
    </Container>
  );
};

Input.defaultProps = {
  type: 'text',
  fontSize: 'body-xsmall',
};

export default Input;
