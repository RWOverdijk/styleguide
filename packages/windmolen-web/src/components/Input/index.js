// @flow
import React, { type Node } from 'react';
import styled from 'styled-components';
import Autocomplete from 'react-autocomplete';
import Base from '../Base';
import Icon from '../Icon';
import { colors } from '../../globals';
import { media } from 'styled-bootstrap-grid';

type InputProps = {
  error?: Node,
  touched?: boolean,
  isValid?: boolean,
  placeholderRight?: boolean,
  label?: Node,
  onIconClick?: Function,
  type?: string,

  /** The name of the icon. */
  icon?: string,

  /** CSS class for the container around `<input />`. */
  className?: string
};

const getInputState = (props: InputProps) => {
  if (!props.touched) {
    return colors.transparent;
  }

  if (props.isValid) {
    return colors.green;
  }

  if (props.error || props.isValid === false) {
    return colors.red;
  }
};

const StyledInput = Base.withComponent('input').extend`
  background-color: ${colors.alabaster};
  border: 0;
  box-shadow: inset 0 0 5px 0 rgba(0, 0, 0, 0.12);
  color: ${props => props.disabled ? colors.silver : colors.charcoalGray};
  display: block;
  font-size: 16px;
  outline: 0;
  padding: 9px 20px;
  padding-right: ${props => props.icon ? '70px' : '20px'};
  width: 100%;
  height: 50px;
  position: relative;
  z-index: 2;
  ${props => props.touched && !props.suggestions ? `border-bottom: 1px solid ${getInputState(props)};` : ''}

  &:active,
  &:focus {
    outline: 0;
    ${props => !props.suggestions && `border-bottom: 1px solid ${!props.disabled ? colors.charcoalGray : colors.transparent};`}
  }

  &::placeholder {
    color: ${props => props.disabled ? colors.silver : colors.warmGray};
    text-align: ${props => props.placeholderRight ? 'right' : 'left'}
  }
`;

const Container = Base.extend`
  position: relative;
`;

const StyledLabel = Base.extend`
  font-size: 16px;
  color: ${colors.charcoalGray};
  margin-bottom: 5px;

  ${media.desktop`
    font-size: 20px;
    margin-bottom: 10px;
  `}
`;

const StyledErrorMessage = Base.extend`
  font-size: 14px;
  color: ${colors.red};
  line-height: 2;
`;

const StyledIcon = styled(Icon)`
  position: absolute;
  right: 5px;
  top: 5px;
`;

const StyledIconContainer = Base.extend`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  width: 50px;
  background-color: ${colors.charcoalGray};
  z-index: 3;

  &:focus,
  &:active,
  &:hover {
    background-color: ${colors.shuttleGray};
  }

  &:hover {
    cursor: pointer;
  }
`;

const StyledAutoSuggestion = Base.extend`
  position: absolute;
  box-shadow: 0 11px 15px 0 rgba(0, 0, 0, 0.1);
  top: 0;
  left: 0;
  right: 0;
  z-index: 1;
  padding-top: 55px;
  padding-bottom: 5px;

  .suggestion {
    padding-left: 20px;
    padding-right: 20px;
    margin-top: 5px;
    margin-bottom: 5px;
    font-size: 16px;
    line-height: 32px;

    &:focus,
    &:active,
    &:hover {
      color: ${colors.silver};
    }

    &:hover {
      cursor: pointer;
    }
  }
`;

const Input = ({ className, autoCompleteProps, ...props }: InputProps) => {
  return (
    <Container className={className}>
      {props.label && (
        <StyledLabel>{props.label}</StyledLabel>
      )}

      {props.suggestions ? (
        <Autocomplete
          items={props.suggestions}
          wrapperStyle={{  }}
          renderInput={({ ref, value, ...inputProps }) => (
            <StyledInput
              innerRef={(node) => ref(node)}
              {...inputProps}
              {...props}
            />
          )}
          renderMenu={(children) => (
            <StyledAutoSuggestion>
              {children}
            </StyledAutoSuggestion>
          )}
          {...autoCompleteProps}
        />
      ) : (
        <StyledInput {...props} />
      )}

      {props.error && (
        <StyledErrorMessage>{props.error}</StyledErrorMessage>
      )}

      {props.icon && (
        <StyledIconContainer onClick={props.onIconClick}>
          <StyledIcon variant={1} name={props.icon} fontSize="40px" />
        </StyledIconContainer>
      )}
    </Container>
  );
};

Input.defaultProps = {
  type: 'text'
};

export default Input;
