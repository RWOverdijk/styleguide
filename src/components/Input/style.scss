@import "../../css/_form-globals.scss";

.Van-Text-Input {
  background-color: $input-bg;
  border: 0;
  box-sizing: border-box;
  color: $input-color;
  display: block;
  font-size: $input-font-size;
  padding: $input-padding;
  width: 100%;
}
