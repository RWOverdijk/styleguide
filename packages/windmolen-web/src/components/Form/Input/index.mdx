---
name: Input
menu: Forms
route: /Forms/Input
---

import { Playground, PropsTable } from 'docz'
import classNames from 'classnames';
import Wrapper from '../../Wrapper';
import Input from './';

# Input

## Props
<PropsTable of={Input} />

## Basic usage
<Playground wrapper={Wrapper}>
  <Input />
</Playground>

## With label
<Playground wrapper={Wrapper}>
  <Input label="Input with label" id="test" />
</Playground>

## Valid input
<Playground wrapper={Wrapper}>
  <Input touched={true} isValid={true} />
</Playground>

## Input with error
<Playground wrapper={Wrapper}>
  <Input touched={true} isValid={false} />
</Playground>

<Playground wrapper={Wrapper}>
  <Input touched={true} error="Oops. You messed up." />
</Playground>

## Disabled
<Playground wrapper={Wrapper}>
  <Input disabled value="Disabled input" />
</Playground>

## With placeholder
<Playground wrapper={Wrapper}>
  <Input placeholder="Type on me" />
</Playground>

## With placeholder to the right
<Playground wrapper={Wrapper}>
  <Input placeholder="kg" placeholderRight />
</Playground>

## With placeholderRight always visible
<Playground wrapper={Wrapper}>
  <Input placeholder="kWh" placeholderRight placeholderAlwaysVisible />
</Playground>

## With icon
<Playground wrapper={Wrapper}>
  <Input icon="search" iconVariant="alternate" label="Input with icon" />
</Playground>

## With autosuggestions
The module https://github.com/reactjs/react-autocomplete is being used for the
autocompletion. Its props can be overwritten using the `autoCompleteProps` prop
as shown in the example below.
<Playground wrapper={Wrapper}>
  <Input
    icon="search"
    suggestions={[
      { label: 'apple',  id: 1 },
      { label: 'banana', id: 2 },
      { label: 'pear',   id: 3 },
      { label: 'apple 1',  id: 4 },
      { label: 'banana 2', id: 5 },
      { label: 'pear 3',   id: 6 },
      { label: 'apple 4',  id: 7 },
      { label: 'banana 5', id: 8 },
      { label: 'pear 6',   id: 9 },
    ]}
    autoCompleteProps={{
      getItemValue: (item) => item.label,
      renderItem: (item, isHighlighted) => {
        // Use the 'suggestion' class to inherit the styleguide's style.
        const suggestionClasses = classNames('suggestion', {
          highlighted: isHighlighted,
        })
        return (
          <div className={suggestionClasses} key={item.id}>
            {item.label}
          </div>
        );
      },
    }}
  />
  <br/>
  <br/>
  <br/>
  <br/>
  <br/>
  <br/>
  <br/>
  <br/>
</Playground>
