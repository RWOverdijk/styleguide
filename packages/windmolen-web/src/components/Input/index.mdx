---
name: Input
menu: Controls
route: /Controls/Input
---

import { Playground, PropsTable } from 'docz'
import windmill from '../../assets/icons/windmill.svg';
import Wrapper from '../Wrapper';
import Input from './';

# Input

## Props
<PropsTable of={Input} />

## Basic usage
<Playground>
  <Input />
</Playground>

## With label
<Playground>
  <Input label="Input with label" />
</Playground>

## Valid input
<Playground>
  <Input touched={true} isValid={true} />
</Playground>

## Input with error
<Playground>
  <Input touched={true} isValid={false} />
</Playground>

<Playground>
  <Input touched={true} error="Oops. You messed up." />
</Playground>

## Disabled
<Playground>
  <Input disabled value="Disabled input" />
</Playground>

## With placeholder
<Playground>
  <Input placeholder="Type on me" />
</Playground>

## With placeholder to the right
<Playground>
  <Input placeholder="kg" placeholderRight />
</Playground>

## With icon
<Playground>
  <Input withIcon={windmill} />
</Playground>
