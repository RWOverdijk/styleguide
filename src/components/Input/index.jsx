import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import './style.scss';

const Input = ({
  meta,
  input,
  className,
  type,
  ...props
}) => {
  const { error, warning, touched } = meta;
  const classNames = classnames(
    'Van-TextInput',
    className,
    { 'Van-TextInput--error': touched && error },
    { 'Van-TextInput--warning': touched && warning },
    { 'Van-TextInput--success': touched && !error && !warning },
  );

  return (
    <div className={classNames}>
      <input
        className="Van-TextInput-input"
        id={input.name || props.id}
        type={type}
        {...props}
        {...input}
      />

      {touched && typeof error === 'string' && <div className="Van-TextInput-error-message">{error}</div>}
    </div>
  );
};

Input.propTypes = {
  meta: PropTypes.object,
  input: PropTypes.object,
  className: PropTypes.string,
  id: PropTypes.string,
  type: PropTypes.string,
};

Input.defaultProps = {
  meta: {},
  input: {},
  className: '',
  id: '',
  type: 'text',
};

export default Input;
