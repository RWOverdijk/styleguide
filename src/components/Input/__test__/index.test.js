import React from 'react';
import renderer from 'react-test-renderer';
import Input from '../';

describe('<Input/>', () => {
  // it("renders input text", () => {
  //   const wrapper = render(<Input meta={{error:null}} type="text"/>)

  //   const input = wrapper.find('input');
  //   expect(input.length).toEqual(1);
  //   expect(input[0].attribs.type).toBe("text");
  //   expect(input[0].attribs.class).toBe('Van-Text-Input');
  // });

  // it("contains informed className", () => {
  //   const wrapper = render(<Input meta={{error:null}} type="text" className= "someClassName"/>)

  //   const input = wrapper.find('input');
  //   expect(input[0].attribs.class).toBe('Van-Text-Input someClassName');
  // });

  it("renders with specific className", () => {
    const tree = renderer.create(
      <Input className="my-class" />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders with specific attributes", () => {
    const tree = renderer.create(
      <Input data-id="0" />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
