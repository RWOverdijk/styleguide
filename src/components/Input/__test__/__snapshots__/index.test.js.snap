// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input/> renders with specific attributes 1`] = `
<div
  className="Van-TextInput"
>
  <input
    className="Van-TextInput-input"
    data-id="0"
    id=""
    type="text"
  />
</div>
`;

exports[`<Input/> renders with specific className 1`] = `
<div
  className="Van-TextInput my-class"
>
  <input
    className="Van-TextInput-input"
    id=""
    type="text"
  />
</div>
`;
